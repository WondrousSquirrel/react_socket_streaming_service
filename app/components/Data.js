import React, {useRef, useEffect} from 'react';
import PropTypes from 'prop-types';
import { faArrowUp, faArrowDown } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

function usePrevious(value) {
    const ref = useRef();
    useEffect(() => {
        ref.current = value;
    }, [value]);
    return ref.current;
}

const Data = ({data}) => {
    const prevPrice = data ? usePrevious(data.price) : 0;
    const priceDiff = prevPrice === 0 ? 0 : data.price - prevPrice;
    return (
        data ?
            <div className="card">
                <div className="card-body">
                    <table>
                        <thead>
                            <tr>
                                <th scope="col">Ticker</th>
                                <th scope="col">Exchange</th>
                                <th scope="col">Price</th>
                                <th scope="col">Change</th>
                                <th scope="col">Change Percent</th>
                                <th scope="col">Last Trade Time</th>
                                <th scope="col">Dividend</th>
                                <th scope="col">Yield</th>
                            </tr>
                        </thead>
                        <tbody>
                            <tr>
                                <th scope="row">{data.ticker}</th>
                                <th scope="row">{data.exchange}</th>
                                <th scope="row">{data.price}</th>
                                <th scope="row">{data.change}</th>
                                <th scope="row">{data.change_percent}</th>
                                <th scope="row">{data.last_trade_time}</th>
                                <th scope="row">{data.dividend}</th>
                                <th scope="row">{data.yield}</th>
                            </tr>
                        </tbody>
                    </table>
                </div>
                <div className="diff-info">
                    <p className={priceDiff > 0 ? 'green-text' : 'red-text'}>{priceDiff.toFixed(2)}</p>
                    {priceDiff > 0 ? <FontAwesomeIcon icon={faArrowUp} /> : <FontAwesomeIcon icon={faArrowDown} />}
                </div>
            </div>
            : <p>Loading...</p>
    );
};

Data.propTypes = {
    data: PropTypes.object,
};

export default Data;
