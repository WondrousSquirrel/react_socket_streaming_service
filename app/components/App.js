import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import io from 'socket.io-client';
import PropTypes from 'prop-types';
import Data from './Data';

import '../styles/application.scss';
import { recieveDataAction } from '../actions/action';

const App = ({recieveData, data}) => {
    let socket;
    const stockSymbol = 'AAPL';
    /* fixes warning in test case: Cannot read property '_cookieJar' of null*/
    if (process.env.NODE_ENV !== 'test')  socket = io('http://localhost:4000');
    useEffect(() => {
        if(process.env.NODE_ENV !== 'test') {
            socket.on('connect', () => {
                console.log('connected to ws');
                socket.on(stockSymbol, (_data) => {
                    recieveData(_data);
                });
                socket.emit('ticker', stockSymbol);
            });
            return () => {
                socket.on('disconnect', () => {
                    console.log('disconnected');
                });
            };
        }
        return () => null;
    }, []);
    return (
        <div className="stock-ticker">
            <h1>Price Ticker</h1>
            <Data data={data} />
        </div>
    );
};

const mapStateToProps = state => ({
    data: state.ticker.data
});

const mapDispatchToProps = dispatch => ({
    recieveData: (data) => dispatch(recieveDataAction(data))
});

App.propTypes = {
    data: PropTypes.object,
    recieveData: PropTypes.func
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(App);
