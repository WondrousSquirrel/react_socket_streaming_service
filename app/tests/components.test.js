import React from 'react';
import { configure } from 'enzyme';
import App from '../components/App';
import Data from '../components/Data';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
// import { render, screen } from './testUtils';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

const mockStore = configureStore();

configure({ adapter: new Adapter() });

const data = {
    ticker: 'AAPL',
    exchange: 'NASDAQ',
    price: '109.79',
    change: '161.21',
    change_percent: '0.55',
    last_trade_time: '2020-11-10T12:53:18.000Z',
    dividend: '0.54',
    yield: '1.86',
};
const initialState = {
    ticker: { data: data},
};
const store = mockStore(initialState);

describe('Testing Components', () => {
    test('Data snapshot', () => {
        const tree = renderer.create(<Data data={data} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Renders the connected app with initialState', () => {
        const tree = renderer.create(<Provider store={store}><App  /> </Provider>).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
