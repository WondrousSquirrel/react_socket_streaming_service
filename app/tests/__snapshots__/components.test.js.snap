// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Components Data snapshot 1`] = `
<div
  className="card"
>
  <div
    className="card-body"
  >
    <table>
      <thead>
        <tr>
          <th
            scope="col"
          >
            Ticker
          </th>
          <th
            scope="col"
          >
            Exchange
          </th>
          <th
            scope="col"
          >
            Price
          </th>
          <th
            scope="col"
          >
            Change
          </th>
          <th
            scope="col"
          >
            Change Percent
          </th>
          <th
            scope="col"
          >
            Last Trade Time
          </th>
          <th
            scope="col"
          >
            Dividend
          </th>
          <th
            scope="col"
          >
            Yield
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th
            scope="row"
          >
            AAPL
          </th>
          <th
            scope="row"
          >
            NASDAQ
          </th>
          <th
            scope="row"
          >
            109.79
          </th>
          <th
            scope="row"
          >
            161.21
          </th>
          <th
            scope="row"
          >
            0.55
          </th>
          <th
            scope="row"
          >
            2020-11-10T12:53:18.000Z
          </th>
          <th
            scope="row"
          >
            0.54
          </th>
          <th
            scope="row"
          >
            1.86
          </th>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="diff-info"
  >
    <p
      className="red-text"
    >
      NaN
    </p>
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-arrow-down fa-w-14 "
      data-icon="arrow-down"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </div>
</div>
`;

exports[`Testing Components Renders the connected app with initialState 1`] = `
Array [
  <div
    className="stock-ticker"
  >
    <h1>
      Price Ticker
    </h1>
    <div
      className="card"
    >
      <div
        className="card-body"
      >
        <table>
          <thead>
            <tr>
              <th
                scope="col"
              >
                Ticker
              </th>
              <th
                scope="col"
              >
                Exchange
              </th>
              <th
                scope="col"
              >
                Price
              </th>
              <th
                scope="col"
              >
                Change
              </th>
              <th
                scope="col"
              >
                Change Percent
              </th>
              <th
                scope="col"
              >
                Last Trade Time
              </th>
              <th
                scope="col"
              >
                Dividend
              </th>
              <th
                scope="col"
              >
                Yield
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <th
                scope="row"
              >
                AAPL
              </th>
              <th
                scope="row"
              >
                NASDAQ
              </th>
              <th
                scope="row"
              >
                109.79
              </th>
              <th
                scope="row"
              >
                161.21
              </th>
              <th
                scope="row"
              >
                0.55
              </th>
              <th
                scope="row"
              >
                2020-11-10T12:53:18.000Z
              </th>
              <th
                scope="row"
              >
                0.54
              </th>
              <th
                scope="row"
              >
                1.86
              </th>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        className="diff-info"
      >
        <p
          className="red-text"
        >
          NaN
        </p>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-arrow-down fa-w-14 "
          data-icon="arrow-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
    </div>
  </div>,
  " ",
]
`;
